### PR TITLE
S3 Repository: Add back repository level credentials

### PR DESCRIPTION
**Pull request by rjernst**
Specifying s3 access and secret keys inside repository settings are not
secure. However, until there is a way to dynamically update secure
settings, this is the only way to dynamically add repositories with
credentials that are not known at node startup time. This commit adds
back `access_key` and `secret_key` s3 repository settings, but protects
it with a required system property `allow_insecure_settings`.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java [hunk 1]
```diff
@@ -24,6 +24,8 @@
 import java.util.EnumSet;
 import java.util.Set;
 
+import org.elasticsearch.cluster.routing.IllegalShardRoutingStateException;
```
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java]: Not needed?

File: core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java [hunk 2]
```diff
@@ -134,6 +136,23 @@ SecureString getFallback(Settings settings) {
     }
 
     /**
+     * A setting which contains a sensitive string, but which for legacy reasons must be found outside secure settings.
+     * @see #secureString(String, Setting, Property...)
+     */
+    public static Setting<SecureString> inecureString(String name) {
```
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java]: Wrong name for the method?

**Review by dadoonet — CHANGES_REQUESTED** [review]
Left minor comments.
I think we may be need to warn when we start the node if `allow_insecure_settings` is set, no?

**Comment by rjernst** [comment]
> I think we may be need to warn when we start the node if allow_insecure_settings is set

For what purpose? Every insecure setting is automatically marked as deprecated, so simply using the setting will cause a deprecation warning.

File: core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java (hunk 2 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java]: Ooops, thanks! Fixed in ad4c625

File: core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java (hunk 1 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java]: Indeed, removed in ad4c625

**Comment by dadoonet** [comment]
> Every insecure setting is automatically marked as deprecated, so simply using the setting will cause a deprecation warning.

Ha that's right. But as it's up to the user to use that insecure setting when he registers a new repository, the admin guy who actually starts elasticsearch server might not be aware at startup time that someone else might use an insecure setting at some point.

**Review by dadoonet — APPROVED** [review]

**Comment by s1monw** [comment]
@rjernst what happens if somebody has a insecure setting but then goes and sets the property to false will the node start? And if the node has started and recovers a cluster state with such a setting will it fail?

**Comment by rjernst** [comment]
For both questions, constructing the S3Repository would fail. However, since the s3 repositories are not created within Node construction, I don't think the node will fail, just that repository would be unavailable.

**Comment by s1monw** [comment]
@rjernst LGTM then

File: core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java [hunk 3]
```diff
@@ -24,25 +24,25 @@
...
-
 /**
  * A secure setting.
  *
  * This class allows access to settings from the Elasticsearch keystore.
  */
 public abstract class SecureSetting<T> extends Setting<T> {
+
+    /** Determines whether legacy settings with sensitive values should be allowed. */
+    private static final boolean ALLOW_INSECURE_SETTINGS = Booleans.parseBoolean(System.getProperty("allow_insecure_settings", "false"));
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/settings/SecureSetting.java]: I think this should be pre-fixed with `es.` like we try to do with other custom system properties that we use. I'm not sure if we're 100% consistent, but we should aim to be when adding new ones.

**Review by jasontedor — DISMISSED** [review]
I didn't do a full review, I just left one comment about the name of the system property. We have somewhat of a convention that we prefix these with `es.` that I think we should follow here.

**Pull request merged**